uamqp/aio/_connection_async.py

```diff
@@ -376,17 +440,30 @@ async def _listen_one_frame(self, **kwargs):
         #    raise Exception("Stop")  # TODO: Stop listening
 
     async def listen(self, wait=False, batch=1, **kwargs):
-        if self.state == ConnectionState.END:
-            raise ValueError("Connection closed.")
-        async with create_task_group() as tg:
+        try:
+            raise self._error
+        except TypeError:
+            pass
+        try:
+            if self.state == ConnectionState.END:
+                raise ValueError("Connection closed.")
+            futures = []
             for _ in range(batch):
-                await tg.spawn(self._listen_one_frame, **kwargs)  # TODO: Close on first exception
-
-
-        if self.state not in _CLOSING_STATES:
-            now = time.time()
-            if self._get_local_timeout(now) or (await self._get_remote_timeout(now)):
-                await self.close(error=None, wait=False)
+                futures.append(asyncio.create_task(self._listen_one_frame(**kwargs)))
+            await asyncio.gather(*futures)   # TODO: Close on first exception
```

Comment:
Another comprehension here:
`await asyncio.gather(*[self._listen_one_frame(**kwargs) for _ in range(batch)])`

Reply:
updated